File: src/generics/game4/game_java/Team.java
package generics.game4.game_java;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class Team <T extends Participant> {
    private String name;
    private List<T> participants = new ArrayList<>();

    public Team(String name) {
        this.name = name;
    }

    public void addNewParticipant(T participant) {
        participants.add(participant);
        System.out.println("В комманду \"" + name + "\" добавлен участник " + participant.getName()
        + ", возраст: " + participant.getAge());
    }

    public void playWith(Team<T> team) {
        Random random = new Random();
        int i = random.nextInt(2);
        String winner = i == 0 ? this.name : team.name;
        System.out.println("Победила комманда \"" + winner + "\"");
    }
}

File: src/generics/game4/Team.kt
package generics.game4

import kotlin.random.Random

class Team<T:Participant>(val name: String) {
    private val participants: MutableList<T> = mutableListOf()

    fun addNewParticipant(participant: T) {
        participants.add(participant)
        println("In team $name added a member ${participant.name}")
    }

    fun playWith(team: Team<T>) {
        val i = Random.Default.nextInt(2)
        val winnerName = if (i == 0) name else team.name
        println("Team $winnerName has won")
    }
}